The cnetdiff application has been refactored to be callable; Makefile tests have been removed and replaced by gtests.

kledmundson opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail including motivation and any context -->
The cnetdiff application has been refactored to be callable; Makefile tests have been removed and replaced by gtests.

## Related Issue
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->https://github.com/DOI-USGS/ISIS3/issues/5322

## How Has This Been Validated?
<!--- All changes need to be validated to confirm that they produce -->
<!--- scientifically accurate results. -->
<!--- If your changes include any new algorithms or changes to existing algorithms, -->
<!--- please indicate any publications or references for them. -->
<!--- If you manually validated the changes please indicate -->
<!--- what data you used and how you checked the results. -->
ISIS make file tests have been replaced with the following gtests.

CnetdiffReportFirst
CnetdiffReportFull

Results of ctest suite...

ctest -R Cnetdiff --output-on-failure
Test project
    Start 1798:CompareNetsReportFirstDiff.FunctionalTestCnetdiffReportFirst
1/2 Test #1798: CompareNetsReportFirstDiff.FunctionalTestCnetdiffReportFirst ...   Passed    0.74 sec
Start 1799: CompareNetsReportFull.FunctionalTestCnetdiffReportFull
2/2 Test #1799:
CompareNetsReportFull.FunctionalTestCnetdiffReportFull .........   Passed    0.56 sec

100% tests passed, 0 tests failed out of 2
Total Test time (real) =   1.50 sec

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Infrastructure change (changes to things like CI or the build system that do not impact users)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [x] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

kledmundson committed:
The cnetedit applciation has been refactored to be callable; Makefile tests have been removed and replace by gtests. Addresses #5322.

acpaquette commented on CHANGELOG.md:
```diff
@@ -45,7 +45,7 @@ release.
 
 ### Added
-- Added rclone to run dependencies in meta.yaml [#5183](https://github.com/DOI-USGS/ISIS3/issues/5183)
+- Cnetedit has been refactored to be callable; old Makefile tests have been removed and replaced by gtests. Issue: [#5322](https://github.com/USGS-Astrogeology/ISIS3/issues/5322), 
```
Can this be moved to changed? If there is no `Changed` section feel free to add a new header titled `Changed`

kledmundson replied:
Done.

acpaquette commented on isis/src/control/apps/cnetdiff/cnetdiff.cpp:
```diff
@@ -0,0 +1,380 @@
+        differences += PvlKeyword("Reason", differenceReason);
+      }
+
+      log.addGroup(differences);
```
Could any addition to the log be changed to `log.addLogGroup`. This function logs to the app at the same time as the group is added to the keyword. With that you can remove the logging loop in the application `main.cpp`

kledmundson replied:
Also done.

acpaquette commented on isis/src/control/objs/ControlNetDiff/ControlNetDiff.cpp:
```diff
@@ -88,14 +87,31 @@ namespace Isis {
    * @return Pvl The collection of all differences
    */
   Pvl ControlNetDiff::compare(FileName &net1Name, FileName &net2Name) {
+
+    ControlNetVersioner cnv1(net1Name);
+    ControlNetVersioner cnv2(net2Name);
+
+    return compare(cnv1, cnv2, net1Name, net2Name);
+  }
+
+
+  /**
+   * Compare two Control Networks in ControNetVersioner form, and return their
+   * differences.
+   *
+   * @param cnv1 first Control Network
+   * @param cnv2 second Control Network
+   * @param net1Name FileName of the first Control Network
+   * @param net2Name FileName of the second Control Network
+   * @return Pvl The collection of all differences
+   */
+  Pvl ControlNetDiff::compare(ControlNetVersioner &cnv1, ControlNetVersioner &cnv2,
+                              FileName &net1Name, FileName &net2Name) {
```
I don't really understand the necessity for this change. What does exposing the new function do that the old signature prevented?

kledmundson replied:
You're right. Major brain lock on my part. Reverted back to the original ControlNetDiff.cpp;h.

acpaquette reviewed (commented):
Overall looks great! Couple small things to address

kledmundson committed:
Moved CHANGELOG.md entry for cnetdiff changes from "Added" section to "Changed" section per reviewer suggestion. Addresses #5322.

kledmundson committed:
Reverted back to original ControlNetDiff source and header files per reviewer suggestion. Addresses #5322.

kledmundson committed:
Changes to cnetdiff.cpp and main.cpp to use the addLogGroup method to add PvlGroups to Pvl. Per reviewer suggestion. Addresses #5322.

kledmundson commented:
Thanks for the very quick review Adam.
I've made the changes and have hopefully addressed your suggestions.

Ken

On Wed, Nov 1, 2023 at 12:57 AM acpaquette ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Overall looks great! Couple small things to address
> ------------------------------
>
> In CHANGELOG.md
> <https://github.com/DOI-USGS/ISIS3/pull/5323#discussion_r1378485280>:
>
> > @@ -45,7 +45,7 @@ release.
>    outputsuffix, both, or neither for naming convention purposes. [#5162](https://github.com/DOI-USGS/ISIS3/pull/5162)
>
>  ### Added
> -- Added rclone to run dependencies in meta.yaml [#5183](https://github.com/DOI-USGS/ISIS3/issues/5183)
> +- Cnetedit has been refactored to be callable; old Makefile tests have been removed and replaced by gtests. Issue: [#5322](https://github.com/USGS-Astrogeology/ISIS3/issues/5322),
>
> Can this be moved to changed? If there is no Changed section feel free to
> add a new header titled Changed
> ------------------------------
>
> In isis/src/control/apps/cnetdiff/cnetdiff.cpp
> <https://github.com/DOI-USGS/ISIS3/pull/5323#discussion_r1378488033>:
>
> > +        ignorekeys += PvlKeyword("DateTime", "true");
> +      }
> +
> +      // compare ControlNetVersioners
> +      Compare(cnv1, cnv2);
> +
> +      PvlGroup differences("Results");
> +      if (filesMatch) {
> +        differences += PvlKeyword("Compare", "Identical");
> +      }
> +      else {
> +        differences += PvlKeyword("Compare", "Different");
> +        differences += PvlKeyword("Reason", differenceReason);
> +      }
> +
> +      log.addGroup(differences);
>
> Could any addition to the log be changed to log.addLogGroup. This
> function logs to the app at the same time as the group is added to the
> keyword. With that you can remove the logging loop in the application
> main.cpp
> ------------------------------
>
> In isis/src/control/objs/ControlNetDiff/ControlNetDiff.cpp
> <https://github.com/DOI-USGS/ISIS3/pull/5323#discussion_r1378491846>:
>
> > +
> +    ControlNetVersioner cnv1(net1Name);
> +    ControlNetVersioner cnv2(net2Name);
> +
> +    return compare(cnv1, cnv2, net1Name, net2Name);
> +  }
> +
> +
> +  /**
> +   * Compare two Control Networks in ControNetVersioner form, and return their
> +   * differences.
> +   *
> +   * @param cnv1 first Control Network
> +   * @param cnv2 second Control Network
> +   * @param net1Name FileName of the first Control Network
> +   * @param net2Name FileName of the second Control Network
> +   * @return Pvl The collection of all differences
> +   */
> +  Pvl ControlNetDiff::compare(ControlNetVersioner &cnv1, ControlNetVersioner &cnv2,
> +                              FileName &net1Name, FileName &net2Name) {
>
> I don't really understand the necessity for this change. What does
> exposing the new function do that the old signature prevented?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DOI-USGS/ISIS3/pull/5323#pullrequestreview-1707718918>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUGSUUDCAHRDTH2BF7JMHLYCH6E7AVCNFSM6AAAAAA6XXKFPKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTOMBXG4YTQOJRHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

acpaquette commented on CHANGELOG.md:
```diff
@@ -45,7 +46,6 @@ release.
   outputsuffix, both, or neither for naming convention purposes. [#5162](https://github.com/DOI-USGS/ISIS3/pull/5162)
 
 ### Added
-- Added rclone to run dependencies in meta.yaml [#5183](https://github.com/DOI-USGS/ISIS3/issues/5183)
```
@kledmundson One last thing, it looks like this entry got erroneously removed. Could we add it back in?

kledmundson replied:
Done.

kledmundson committed:
Put back line erroneously removed in CHANGELOG.md, per review. Addresses #5322.
